dev/NumberBox/NumberBox.cpp

```diff
@@ -0,0 +1,312 @@
+    m_TextBox.KeyUp({ this, &NumberBox::OnNumberBoxKeyUp });
+    this->PointerWheelChanged({ this, &NumberBox::OnScroll });
+
+    // Initializations - Tools, etc.
+    Formatter = winrt::DecimalFormatter();
+    IRounder = winrt::IncrementNumberRounder();
+    SRounder = winrt::SignificantDigitsNumberRounder();
+    UpdateFormatter();
+    UpdateRounder();
+    // Set Text to reflect preset Value
+    if (s_ValueProperty != DEFAULTVALUE && m_TextBox) {
+        UpdateTextToValue();
+    }
+
+    // Register LostFocus Event
+    if ( m_TextBox ) {
+        m_TextBox.LostFocus({ this, &NumberBox::OnTextBoxLostFocus });
+    }
+}
+
+void  NumberBox::OnPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    winrt::IDependencyProperty property = args.Property();
+    
+    // Update visual state for spin buttons if placement mode changed
+    if (property == s_SpinButtonPlacementModeProperty)
+    {
+        SetSpinButtonVisualState();
+    }
+    else if (property == s_IntegerDigitsProperty || property == s_FractionDigitsProperty || property == s_SignificantDigitsProperty || property == s_IsDecimalPointAlwaysDisplayedProperty || property == s_IsZeroSignedProperty)
+    {
+        UpdateFormatter();
+    }
+    else if (property == s_RoundingAlgorithmProperty || property == s_NumberRounderProperty || property == s_SignificantDigitPrecisionProperty || property == s_IncrementPrecisionProperty)
+    {
+        UpdateRounder();
+    }
+    else if (property == s_HeaderProperty)
+    {
+        SetHeader();
+    }
+
+}
+
+// Trigger any validation, rounding, and processing done onLostFocus
+void NumberBox::OnTextBoxLostFocus(winrt::IInspectable const& sender, winrt::RoutedEventArgs const& args)
+{
+    ValidateInput();
+}
+
+// Performs all validation steps on input given in textbox. Runs on LoseFocus and stepping.
+void NumberBox::ValidateInput()
+{
+    string InputAsString = winrt::to_string(m_TextBox.Text());
+    // Handles Empty TextBox Case, current behavior is to set Value to default (0)
+    if (InputAsString == "")
+    {
+        this->Value(0);
+        HasError = false;
+        return;
+    }
+
+    winrt::IReference<double> parsedNum = Formatter.ParseDouble(m_TextBox.Text());
+    // TODO: Formatter.Format(value)
+
+    if (parsedNum && IsInBounds(parsedNum.Value()) )
+    {
+        SetErrorState(false);
+        Value(parsedNum.Value());
+        UpdateTextToValue();
+    }
+    else
+    {
+        SetErrorState(true);
+    }
+}
+
+// SpinClicks call to decrement or increment, 
+void NumberBox::OnSpinDownClick(winrt::IInspectable const&  sender, winrt::RoutedEventArgs const& args)
+{
+    StepValue(false);
+}
+void NumberBox::OnSpinUpClick(winrt::IInspectable const& sender, winrt::RoutedEventArgs const& args)
+{
+    StepValue(true);
+}
+void NumberBox::OnNumberBoxKeyUp(winrt::IInspectable const& sender, winrt::KeyRoutedEventArgs const& args)
+{
+    switch (args.Key()) {
```

Comment:
>Key [](start = 17, length = 3)

You won't get Gamepad keys from .Key(), you need to use OriginalKey() to see those. But I suspect we'll want to use "engagement" to change the values with gamepad, I don't think you want to get trapped in this control by gamepad? @austin-lamb?

Reply:
Made this swap in my last commit - don't really have the tools/knowhow to test gamepad interactions and it's probably not an urgent priority so I'll leave this open.